DRIVERS-1707: Preemptively cancel in progress operations when SDAM heartbeats timeout. (#1170)
mongodb/specifications: source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst

      /**
       *  Mark all current Connections as stale, clear the WaitQueue, and mark the pool as "paused".
       *  closeInUseConnections specifies whether a pool SHOULD force closing "in use" connections.
       *  No connections may be checked out or created in this pool until ready() is called again.
       */

>       clear(closeInUseConnections: Boolean): void;

Author: Optional with default false?

Reviewer: I think `closeInUseConnections` can be optional, especially for drivers without background thread.
@patrickfreed what do you think?

as soon as possible. The pool MUST NOT rely on WaitQueueTimeoutMS to clear
requests from the WaitQueue.

Additionally, a pool SHOULD force the next maintenance step to remove perished 
connections. The maintenance logic SHOULD ensure that prune step is aware about 
changed pool generation. As part of this request, a pool SHOULD inform maintenance
thread whether "in use" connections should be closed as well. A pool SHOULD be 
informed about it via closeInUseConnections parameter in Clear method.

> 

Reviewer: Probably "a pool SHOULD force the next maintenance step..." needs to be clarified. Does this mean that the next step should be scheduled to run as soon as possible? 
Maybe something like: The next Background Thread Run SHOULD be scheduled as soon as possible. Next pruning iteration MUST close "in use" perished connections if requested by closeInUseConnections flag...

Author: done

Reviewer: Maybe we should omit the last "requested by closeInUseConnections flag." part? As it might be interpreted as scheduling the next run sooner will be done only in `closeInUseConnections:true` case.

Reviewer 2: If we do that, can we bump it to its own paragraph below this one? That way the `closeInUseConnections` and background thread scheduling parts will be in separate paragraphs.

Reviewer: I did not notice the 
`A pool SHOULD allow immediate scheduling of the next background thread iteration after a clear is performed. ` sentence. I am fine with current wording.

nit: remove the dot from `connections.requested`

Author: done

as soon as possible. The pool MUST NOT rely on WaitQueueTimeoutMS to clear
requests from the WaitQueue.

> The clearing method MUST provide the option to close any in-use connections as part

Reviewer: I am wondering whether we should clarify that only `in-use` connections removal should be limited up to the generation at the moment of clear. To prevent the following scheduling (steps 3-5 happen fast):
1. pool.ready() (gen1)
2. pool.clear(inUser:true) (gen2)
3. pool.ready() (gen2)
4. pool.clear(inUse:false) (gen3)
5. pool.ready() (gen3)
6. prune runs and closes inUse with `gen <= gen2` (the correct behaviour is `gen <= gen1`

cc @patrickfreed

Reviewer 2: Yeah this is a really good point. Maybe add the following line:
>The pool MUST only close in use connections whose generation is less than or equal to the generation of the pool at the moment of the clear (before the increment) that used the `closeInUseConnections` flag.

Unfortunately, I don't think there's a non-racy way to unit test this that I can think of.

Author: done

as soon as possible. The pool MUST NOT rely on WaitQueueTimeoutMS to clear
requests from the WaitQueue.

> The clearing method(s) MUST provide a way to close perished connections as part 

Reviewer 2: I think the wording of this paragraph was a bit clearer before, can we revert the most recent changes to it?

as soon as possible. The pool MUST NOT rely on WaitQueueTimeoutMS to clear
requests from the WaitQueue.

The clearing method MUST provide the option to close any in-use connections as part
of the clearing (henceforth referred to as the closeInUseConnections flag in this

> specification). The closing of these connections MUST NOT be done eagerly, and

Reviewer 3: > The closing of these connections MUST NOT be done eagerly

Why? In PyMongo we already close idle connections eagerly and I imagine we would do the same for in use connections too.

Reviewer 2: The intention with this requirement is to avoid blocking the connection pool. Per https://unix.stackexchange.com/questions/145526/can-close-block, it looks like that only happens if `SO_LINGER` is set on the socket though, so maybe this requirement isn't necessary. Perhaps it should be stated more directly as:

>The closing of these connections MUST NOT block the pool or prevent it from processing further requests.

Note that we're also discussing some terminology changes here to make this "The interruption of these connections...", with the actual "closing" being delegated to checkIn.

Reviewer 3: Ooo yeah I like that more direct phrasing better. It makes the intention clear.

Reviewer 2: Cool, we'll update it to that then, cc @DmitryLukyanov

as soon as possible. The pool MUST NOT rely on WaitQueueTimeoutMS to clear
requests from the WaitQueue.

> The clearing method MUST provide the option to interrupt any in-use connections as part
> of the clearing (henceforth referred to as the interruptInUseConnections flag in this
> specification) where by "interrupt" drivers SHOULD mean canceling whatever task the 
> connection is currently performing and marks the connection as perished (e.g. by closing 
> the underlying socket). The interrupting of these connections MUST NOT block the pool or prevent 
> it from processing further requests. The next background thread run SHOULD be scheduled 
> as soon as possible if it's responsible for interrupting these connections requested 
> by interruptInUseConnections flag.
> The pool MUST only interrupt in use connections whose generation is less than or equal 
> to the generation of the pool at the moment of the clear (before the increment) 
> that used the interruptInUseConnections flag. Any operations that have their connections 
> interrupted in this way MUST fail with a retryable error. If possible, the error SHOULD 
> be a PoolClearedError with the following message: "Connection to <pool address> interrupted 
> due to server monitor timeout".
> 

Reviewer 2: ```suggestion
The clearing method MUST provide the option to interrupt any in-use connections as part
of the clearing (henceforth referred to as the interruptInUseConnections flag in this
specification). "Interrupting a Connection" is defined as canceling whatever task the 
Connection is currently performing and marking the Connection as perished (e.g. by closing 
its underlying socket). The interrupting of these Connections MUST be performed as soon as possible
but MUST NOT block the pool or prevent it from processing further requests. If the pool has a background
thread, and it is responsible for interrupting in-use connections, its next run MUST be scheduled as soon as
possible.

The pool MUST only interrupt in-use Connections whose generation is less than or equal 
to the generation of the pool at the moment of the clear (before the increment) 
that used the interruptInUseConnections flag. Any operations that have their Connections 
interrupted in this way MUST fail with a retryable error. If possible, the error SHOULD 
be a PoolClearedError with the following message: "Connection to <pool address> interrupted 
due to server monitor timeout".
```

Reviewer 2: In this suggestion I added the more general requirement that "The interrupting of these Connections MUST be performed as soon as possible", since I think this expresses the immediate-schedule requirement in a way that applies to drivers that don't use a background thread or don't need to use one for this purpose. I made it a "MUST", since I think without expedient interruption, this feature isn't all that useful. As part of that, it made sense to also update the immediate-schedule requirement to a MUST too (I think you may have suggested this initially). Let me know if these changes seem good.

I also added some minor wording updates.

> Clearning a load balanced pool

Reviewer 3: Clearning -> Clearing

Also there needs to be a blank line before this section title.

Author: done

Background Thread
^^^^^^^^^^^^^^^^^

> A Pool SHOULD have a background Thread (also known as maintenance thread) that is responsible for

Reviewer: Maybe introducing additional terminology can be avoided here? So we could stick with just `background` naming.

Author: I would say, it's better to use maintenance naming everywhere.., but I agree, no need to change it now, reverted

thread SHOULD

-  Populate `Connections <#connection>`_ to ensure that the pool always satisfies minPoolSize.
-  Remove and close perished available `Connections <#connection>`_ including "in use" connections if `closeInUseConnections` option was set to true in the most recent pool clear.
- Apply timeouts to connection establishment per `Client Side Operations
  Timeout: Background Connection Pooling
  <../client-side-operations-timeout/client-side-operations-timeout.rst#background-connection-pooling>`__.

> A pool SHOULD allow immediate scheduling of the next background thread iteration. 

Reviewer 2: ```suggestion
A pool SHOULD allow immediate scheduling of the next background thread iteration after a clear is performed. 
```

  Timeout: Background Connection Pooling

>   <../client-side-operations-timeout/client-side-operations-timeout.rst#background-connection-pooling>`__.

Author: done

>   <../client-side-operations-timeout/client-side-operations-timeout.rst#background-connection-pooling>`__.

Reviewer 2: can we include a bit about pool clearing being the source of that option? e.g. "including "in use" connections if `closeInUseConnections` option was set to true in the most recent pool clear". This makes it more obvious where the option was set.

>   <../client-side-operations-timeout/client-side-operations-timeout.rst#background-connection-pooling>`__.

Author: done

> A pool SHOULD allow force scheduling the next background thread iteration to remove perished connections including "in use" connections. 

Reviewer 2: ```suggestion
A pool SHOULD allow immediate scheduling of the next background thread iteration.
```
The "SHOULD" here applies to the scheduling behavior, the work done by the thread during the next run is unchanged.

  Timeout: Background Connection Pooling
  <../client-side-operations-timeout/client-side-operations-timeout.rst#background-connection-pooling>`__.

A pool SHOULD allow ability to force run next maintenance iteration to remove perished connections including "in use" connections. 

> In this case, A pool MAY skip populating connections.

Author: not sure whether this step is necessary? It looks like `minPoolSize` can be handled as before even though they will be available only after pool will become `ready`

Reviewer: 1. Maybe "In this case, A pool MAY skip populating connections." can be omitted, as it's part of pausable behaviour: "Connections are not created in the background to satisfy minPoolSize"
2. "force run next maintenance iteration" ==> "Schedule the next Background Thread Run to run as soon as possible" or similar

Author: done

       * The service id for which the pool was cleared for in load balancing mode.
       * See load balancer specification for more information about this field.
       */
      serviceId: Optional<ObjectId>;

      /**
       * A flag whether the pool forced closing "in use" connections as part of the clear.
      */

>       closeInUseConnections: Optional<Boolean>;

Reviewer 2: We should augment the tests to verify that this flag matches what we expect it to be.

Reviewer 2: See my other comment about this: we should update the existing tests to assert on this flag rather than adding a new one.

Author: The current test and implementation designs don't allow doing it

Author: yeah, If I change a place where we trigger cleared pool event, then it's possible to simplify it, done

       * The service id for which the pool was cleared for in load balancing mode.
       * See load balancer specification for more information about this field.
       */
      serviceId: Optional<ObjectId>;

      /**
       * A flag whether the pool forced closing "in use" connections as part of the clear.
      */

>       closeInUseConnections: Optional<Boolean>;

Reviewer 2: cc @BorisDog 

@DmitryLukyanov and I discussed the naming of this option and how it can lead to confusion in driver implementations, namely that driver authors may try to implement this by "closing" (in the CMAP sense, thereby emitting ConnectionClosedEvents) in use connections directly from the background thread. To keep consistent with the existing Connection lifecycle model, it would instead be better if these connections were marked as perished (e.g. by closing the underlying TCP socket), and then the "closing" would happen in checkIn as is the case for other kinds of network errors currently.

In order to better communicate this requirement, we proposed renaming this option to `interruptInUseConnections` and to add a definition of "interrupt", which would read something like "cancels whatever task the connection is currently performing and marks the connection as perished (e.g. by closing the underlying socket)".

Does this sound good to you?

Author: done

background thread after it encounters an error establishing a connection, but
the ServerDescription for the endpoint was not updated accordingly yet.

Why do we need to request the Background Thread Run in Clear logic?
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
If a SDAM logic has observed a network timeout, we assume that all connections
including "in use" connections are no longer healthy. In some cases connections 
will fail to detect the network timeout fast enough, for example server request 
can hang on OS level in TCP retry loop up to 17 minutes until failing. Therefore 
these connections are proactively closed in the case of SDAM network timeout. 
This logic is cross driver and not dependent on OS behavior and configured 
background thread interval.

Why don't we configure TCP_USER_TIMEOUT?
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Ideally reasonable TCP_USER_TIMEOUT for connection can help with detecting stale 
> connections as alternative for manual requesting Background Thread Run in Clear. 
> Unfortunately this approach doesn't work on Windows. Also configuring timeouts 
> smaller than 20 seconds proved to be impossible in some cases.

Reviewer 2: ```suggestion
Ideally, a reasonable TCP_USER_TIMEOUT can help with detecting stale 
connections as an alternative to `closeInUseConnections` in Clear. 
Unfortunately this approach doesn't work on Windows. Also configuring timeouts 
smaller than 20 seconds proved to be impossible in some cases.
```

Reviewer 2: Also, can we provide a bit more context on why this doesn't work on Windows?

Author: Also, AFAIK at least java also has no way to configure this behavior via built in options

background thread after it encounters an error establishing a connection, but
the ServerDescription for the endpoint was not updated accordingly yet.

Why do we need force running prune maintenance call in Clear logic?
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> If a SDAM logic has noticed a network error, we assume that all connections including

Reviewer: Should "network timeout" be used instead of a general "network error"?

Author: true, done

"in use" connections are no longer healthy. In some cases we can't rely on throwing
exception in the connection itself, for example, after network error, server request
can hang on OS level in TCP retry loop and wait up to 17 minutes until failing.
So, to notify pool about changes on network level asap, a pool forces running 
maintenance logic to ensure that connection state is up to date. This logic is 
cross driver and not dependent on OS behavior and configured maintenance interval.

Why don't we configure TCP_USER_TIMEOUT?

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: minor: `~~~~~~~` length should be same as previous line.

Author: I didn't realize this, done

If a SDAM monitor has observed a network timeout, we assume that all connections
including "in use" connections are no longer healthy. In some cases connections 
will fail to detect the network timeout fast enough. For example, a server request 
can hang at the OS level in TCP retry loop up for 17 minutes before failing. Therefore 
these connections MUST be proactively closed in the case of a server monitor network timeout. 
This logic is cross driver and not dependent on OS behavior and configured 
background thread interval.

Why don't we configure TCP_USER_TIMEOUT?
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Ideally, a reasonable TCP_USER_TIMEOUT can help with detecting stale connections as an 
alternative to `closeInUseConnections` in Clear. 
Unfortunately this approach is platform dependent and not each driver allows easily configuring it.
For example, C# driver can configure this socket option on linux only with target frameworks 
higher or equal to .net 5.0. On macOS, there is no straight equavalent for this option, 
it's possible that we can find some equavalent configuration, but this configuration will also 

> require target frameworks higher than or equal to .net 50. The advantage of using Background Thread to 

Reviewer: typo: `.net 5.0`

Author: done

> manage perished connections is that it will work regardless environment setup.

Reviewer: nit: regardless of?

Author: done

background thread after it encounters an error establishing a connection, but
the ServerDescription for the endpoint was not updated accordingly yet.

> Why do we need to request the Background Thread Run in Clear logic?

Reviewer 2: I think the right question here is actually "Why does the pool need to support closing in use connections as part of its clear logic?". The requesting background thread run is more of a nice-to-have perf improvement than the focus of this question. The answer doesn't seem like it needs changes to me. You could add a line that says requesting an immediate backround thread run will speed up this process.

Author: done

background thread after it encounters an error establishing a connection, but
the ServerDescription for the endpoint was not updated accordingly yet.

> Why does the pool need to support closing in use connections as part of its clear logic?

Reviewer 2: closing -> interrupting

Author: done

background thread after it encounters an error establishing a connection, but
the ServerDescription for the endpoint was not updated accordingly yet.

Why does the pool need to support interrupting in use connections as part of its clear logic?

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: minor: the tilde and the previous line should have the same length?

Author: done
